## Exploring Hacker News Posts
##### N:B. This projects for personal practice only, not intented for commercial use or to mislead someone.

In this project, we'll compare two different types of posts from Hacker News, a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts we'll explore begin with either Ask HN or Show HN.

Users submit Ask HN posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll specifically compare these two types of posts to determine the following:

- Between `Ask HN` and `Show HN`, which one receive more comments on average? Do posts created at a certain time receive more comments on average?

- It should be noted that the data set we're working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.

In [1]:
import csv

In [2]:
open_file = open('hacker_news.csv')

In [3]:
hn = csv.reader(open_file)

In [4]:
hn = list(hn)

In [5]:
hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

## Removing Headers from a List of Lists

In [6]:
hn = hn[1:]

In [7]:
hn[:10]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12'],
 ['10482257',
  'Title II kil

## Extracting Ask HN and Show HN Posts

In [8]:
ask_posts = []
show_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    if title.startswith("ask hn"):
        ask_posts.append(row)
    elif title.startswith("show hn"):
        show_posts.append(row)

In [9]:
show_posts[:2]

[['10627194',
  'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform',
  'https://iot.seeed.cc',
  '26',
  '22',
  'kfihihc',
  '11/25/2015 14:03'],
 ['10646440',
  'Show HN: Something pointless I made',
  'http://dn.ht/picklecat/',
  '747',
  '102',
  'dhotson',
  '11/29/2015 22:46']]

In [10]:
ask_posts[:2]

[['12296411',
  'Ask HN: How to improve my personal website?',
  '',
  '2',
  '6',
  'ahmedbaracat',
  '8/16/2016 9:55'],
 ['10610020',
  'Ask HN: Am I the only one outraged by Twitter shutting down share counts?',
  '',
  '28',
  '29',
  'tkfx',
  '11/22/2015 13:43']]

In [11]:
len(ask_posts)

1744

In [12]:
len(show_posts)

1162

We can see that there are 1744 of `Ash HN` and 1162 `Show HN` posts in our dataset. 

##  Calculating the Average Number of Comments for Ask HN and Show HN Posts

In [13]:
ask_average_comments = 0

for row in ask_posts:
    n_comments = row[4]
    n_comments = int(n_comments)
    ask_average_comments += n_comments

average_ask_comments = ask_average_comments / len(ask_posts)
print(average_ask_comments)

14.038417431192661


In [14]:
show_average_comments = 0

for row in show_posts:
    n_comments = row[4]
    n_comments = int(n_comments)
    show_average_comments += n_comments

average_show_comments = show_average_comments / len(show_posts)
print(average_show_comments)

10.31669535283993


The average comments for the title `Ask HN` is around `14` and the average comments for the title `Show HN` is arounf `10`. We can say that there are more comments on `Ash HN` than `Show HN` on average.

## Finding the Amount of Ask Posts and Comments by Hour Created

Now, we will find the amount of ask posts created during each hour of day, along with the number of comments those posts received.

In [37]:
from datetime import datetime

In [38]:
ask_sorted = []

for row in ask_posts:
    created_time = row[6]
    comments = int(row[4])
    ask_sorted.append([created_time, comments])  
print(ask_sorted[2])

['5/2/2016 10:14', 1]


In [39]:
counts_by_hour = {}
comments_by_hour = {}
for i in ask_sorted:
    date = i[0]
    comments = i[1]
    time = datetime.strptime(date, "%m/%d/%Y %H:%M").strftime("%H")
    
    if time in counts_by_hour:
        counts_by_hour[time] += 1
        comments_by_hour[time] += comments
    else:
        counts_by_hour[time] = 1
        comments_by_hour[time] = comments

In [36]:
print(comments_by_hour)

{'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


## Average number of comments on ask posts by hour

Now we are going to calculate the average amount of comments ask posts created at each hour of the day receive.

In [48]:
average_comments_by_hour = [[round(comments_by_hour[x]/counts_by_hour[x],3), x] for x in counts_by_hour]

In [50]:
print(average_comments_by_hour)

[[5.578, '09'], [14.741, '13'], [13.441, '10'], [13.234, '14'], [16.796, '16'], [7.985, '23'], [9.411, '12'], [11.46, '17'], [38.595, '15'], [16.009, '21'], [21.525, '20'], [23.81, '02'], [13.202, '18'], [7.796, '03'], [10.087, '05'], [10.8, '19'], [11.383, '01'], [6.746, '22'], [10.25, '08'], [7.17, '04'], [8.127, '00'], [9.023, '06'], [7.853, '07'], [11.052, '11']]


## Sorting and Printing Values from a List of Lists

In [69]:
swap_average_by_hour = [[h[0],h[1]] for h in average_comments_by_hour]

In [72]:
print(swap_average_by_hour)

[[5.578, '09'], [14.741, '13'], [13.441, '10'], [13.234, '14'], [16.796, '16'], [7.985, '23'], [9.411, '12'], [11.46, '17'], [38.595, '15'], [16.009, '21'], [21.525, '20'], [23.81, '02'], [13.202, '18'], [7.796, '03'], [10.087, '05'], [10.8, '19'], [11.383, '01'], [6.746, '22'], [10.25, '08'], [7.17, '04'], [8.127, '00'], [9.023, '06'], [7.853, '07'], [11.052, '11']]


In [75]:
sorted_swap = sorted(average_comments_by_hour, reverse=True)
print(sorted_swap)

[[38.595, '15'], [23.81, '02'], [21.525, '20'], [16.796, '16'], [16.009, '21'], [14.741, '13'], [13.441, '10'], [13.234, '14'], [13.202, '18'], [11.46, '17'], [11.383, '01'], [11.052, '11'], [10.8, '19'], [10.25, '08'], [10.087, '05'], [9.411, '12'], [9.023, '06'], [8.127, '00'], [7.985, '23'], [7.853, '07'], [7.796, '03'], [7.17, '04'], [6.746, '22'], [5.578, '09']]


In [76]:
for avg, h in sorted_swap:
    time = datetime.strptime(h, "%H").strftime("%H:%M")
    print("{}: {:.2f} average comments per post".format(time, avg)
         )

15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post
13:00: 14.74 average comments per post
10:00: 13.44 average comments per post
14:00: 13.23 average comments per post
18:00: 13.20 average comments per post
17:00: 11.46 average comments per post
01:00: 11.38 average comments per post
11:00: 11.05 average comments per post
19:00: 10.80 average comments per post
08:00: 10.25 average comments per post
05:00: 10.09 average comments per post
12:00: 9.41 average comments per post
06:00: 9.02 average comments per post
00:00: 8.13 average comments per post
23:00: 7.99 average comments per post
07:00: 7.85 average comments per post
03:00: 7.80 average comments per post
04:00: 7.17 average comments per post
22:00: 6.75 average comments per post
09:00: 5.58 average comments per post


On average 38.59 commnets per post were received at 15:00. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set [documentation](https://www.kaggle.com/hacker-news/hacker-news-posts/home), the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.